platforms/zcu102_arm/images/aarch64-xilinx-linux/usr/share/man/man7/man.7
.\" (C) Copyright 1992-1999 Rickard E. Faith and David A. Wheeler
.\" (faith@cs.unc.edu and dwheeler@ida.org)
.\"
.\" %%%LICENSE_START(VERBATIM)
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\"
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\"
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\" %%%LICENSE_END
.\"
.\" Modified Sun Jul 25 11:06:05 1993 by Rik Faith (faith@cs.unc.edu)
.\" Modified Sat Jun  8 00:39:52 1996 by aeb
.\" Modified Wed Jun 16 23:00:00 1999 by David A. Wheeler (dwheeler@ida.org)
.\" Modified Thu Jul 15 12:43:28 1999 by aeb
.\" Modified Sun Jan  6 18:26:25 2002 by Martin Schulze <joey@infodrom.org>
.\" Modified Tue Jul 27 20:12:02 2004 by Colin Watson <cjwatson@debian.org>
.\" 2007-05-30, mtk: various rewrites and moved much text to new man-pages.7.
.\"
.TH MAN 7 2019-03-06 "Linux" "Linux Programmer's Manual"
.SH NAME
man \- macros to format man pages
.SH SYNOPSIS
.B groff \-Tascii \-man
.I file
\&...
.PP
.B groff \-Tps \-man
.I file
\&...
.PP
.B man
.RI [ section ]
.I title
.SH DESCRIPTION
This manual page explains the
.B "groff an.tmac"
macro package (often called the
.B man
macro package).
This macro package should be used by developers when
writing or porting man pages for Linux.
It is fairly compatible with other
versions of this macro package, so porting man pages should not be a major
problem (exceptions include the NET-2 BSD release, which uses a totally
different macro package called mdoc; see
.BR mdoc (7)).
.PP
Note that NET-2 BSD mdoc man pages can be used with
.B groff
simply by specifying the
.B \-mdoc
option instead of the
.B \-man
option.
Using the
.B \-mandoc
option is, however, recommended, since this will automatically detect which
macro package is in use.
.PP
For conventions that should be employed when writing man pages
for the Linux \fIman-pages\fP package, see
.BR man-pages (7).
.SS Title line
The first command in a man page (after comment lines,
that is, lines that start with \fB.\e"\fP) should be
.PP
.RS
.B \&.TH
.I "title section date source manual"
.RE
.PP
For details of the arguments that should be supplied to the
.B TH
command, see
.BR man-pages (7).
.PP
Note that BSD mdoc-formatted pages begin with the
.B Dd
command, not the
.B TH
command.
.SS Sections
Sections are started with
.B \&.SH
followed by the heading name.
.\" The following doesn't seem to be required (see Debian bug 411303),
.\" If the name contains spaces and appears
.\" on the same line as
.\" .BR \&.SH ,
.\" then place the heading in double quotes.
.PP
The only mandatory heading is NAME, which should be the first section and
be followed on the next line by a one-line description of the program:
.PP
.RS
\&.SH NAME
.br
item \e- description
.RE
.PP
It is extremely important that this format is followed, and that there is a
backslash before the single dash which follows the item name.
This syntax is used by the
.BR mandb (8)
program to create a database of short descriptions for the
.BR whatis (1)
and
.BR apropos (1)
commands.
(See
.BR lexgrog (1)
for further details on the syntax of the NAME section.)
.PP
For a list of other sections that might appear in a manual page, see
.BR man-pages (7).
.SS Fonts
The commands to select the type face are:
.TP 4
.B \&.B
Bold
.TP
.B \&.BI
Bold alternating with italics
(especially useful for function specifications)
.TP
.B \&.BR
Bold alternating with Roman
(especially useful for referring to other
manual pages)
.TP
.B \&.I
Italics
.TP
.B \&.IB
Italics alternating with bold
.TP
.B \&.IR
Italics alternating with Roman
.TP
.B \&.RB
Roman alternating with bold
.TP
.B \&.RI
Roman alternating with italics
.TP
.B \&.SB
Small alternating with bold
.TP
.B \&.SM
Small (useful for acronyms)
.PP
Traditionally, each command can have up to six arguments, but the GNU
implementation removes this limitation (you might still want to limit
yourself to 6 arguments for portability's sake).
Arguments are delimited by spaces.
Double quotes can be used to specify an argument which contains spaces.
All of the arguments will be printed next to each other without
intervening spaces, so that the
.B \&.BR
command can be used to specify a word in bold followed by a mark of
punctuation in Roman.
If no arguments are given, the command is applied to the following line
of text.
.SS Other macros and strings
.PP
Below are other relevant macros and predefined strings.
Unless noted otherwise, all macros
cause a break (end the current line of text).
Many of these macros set or use the "prevailing indent."
The "prevailing indent" value is set by any macro with the parameter
.I i
below;
macros may omit
.I i
in which case the current prevailing indent will be used.
As a result, successive indented paragraphs can use the same indent without
respecifying the indent value.
A normal (nonindented) paragraph resets the prevailing indent value
to its default value (0.5 inches).
By default, a given indent is measured in ens;
try to use ens or ems as units for
indents, since these will automatically adjust to font size changes.
The other key macro definitions are:
.SS Normal paragraphs
.TP 9m
.B \&.LP
Same as
.B \&.PP
(begin a new paragraph).
.TP
.B \&.P
Same as
.B \&.PP
(begin a new paragraph).
.TP
.B \&.PP
Begin a new paragraph and reset prevailing indent.
.SS Relative margin indent
.TP 9m
.BI \&.RS " i"
Start relative margin indent: moves the left margin
.I i
to the right (if
.I i
is omitted, the prevailing indent value is used).
A new prevailing indent is set to 0.5 inches.
As a result, all following paragraph(s) will be
indented until the corresponding
.BR \&.RE .
.TP
.B \&.RE
End relative margin indent and
restores the previous value of the prevailing indent.
.SS Indented paragraph macros
.TP 9m
.BI \&.HP " i"
Begin paragraph with a hanging indent
(the first line of the paragraph is at the left margin of
normal paragraphs, and the rest of the paragraph's lines are indented).
.TP
.BI \&.IP " x i"
Indented paragraph with optional hanging tag.
If the tag
.I x
is omitted, the entire following paragraph is indented by
.IR i .
If the tag
.I x
is provided, it is hung at the left margin
before the following indented paragraph
(this is just like
.B \&.TP
except the tag is included with the command instead of being on the
following line).
If the tag is too long, the text after the tag will be moved down to the
next line (text will not be lost or garbled).
For bulleted lists, use this macro with \e(bu (bullet) or \e(em (em dash)
as the tag, and for numbered lists, use the number or letter followed by
a period as the tag;
this simplifies translation to other formats.
.TP
.BI \&.TP " i"
Begin paragraph with hanging tag.
The tag is given on the next line, but
its results are like those of the
.B \&.IP
command.
.SS Hypertext link macros
.TP
.BI \&.UR " url"
Insert a hypertext link to the URI (URL)
.IR url ,
with all text up to the following
.B \&.UE
macro as the link text.
.TP
.B \&.UE \c
.RI [ trailer ]
Terminate the link text of the preceding
.B \&.UR
macro, with the optional
.I trailer
(if present, usually a closing parenthesis and/or end-of-sentence
punctuation) immediately following.
For non-HTML output devices (e.g.,
.BR "man -Tutf8" ),
the link text is followed by the URL in angle brackets; if there is no
link text, the URL is printed as its own link text, surrounded by angle
brackets.
(Angle brackets may not be available on all output devices.)
For the HTML output device, the link text is hyperlinked to the URL; if
there is no link text, the URL is printed as its own link text.
.PP
These macros have been supported since GNU Troff 1.20 (2009-01-05) and
Heirloom Doctools Troff since 160217 (2016-02-17).
.SS Miscellaneous macros
.TP 9m
.B \&.DT
Reset tabs to default tab values (every 0.5 inches);
does not cause a break.
.TP
.BI \&.PD " d"
Set inter-paragraph vertical distance to d
(if omitted, d=0.4v);
does not cause a break.
.TP
.BI \&.SS " t"
Subheading
.I t
(like
.BR \&.SH ,
but used for a subsection inside a section).
.SS Predefined strings
The
.B man
package has the following predefined strings:
.IP \e*R
Registration Symbol: \*R
.IP \e*S
Change to default font size
.IP \e*(Tm
Trademark Symbol: \*(Tm
.IP \e*(lq
Left angled double quote: \*(lq
.IP \e*(rq
Right angled double quote: \*(rq
.SS Safe subset
Although technically
.B man
is a troff macro package, in reality a large number of other tools
process man page files that don't implement all of troff's abilities.
Thus, it's best to avoid some of troff's more exotic abilities
where possible to permit these other tools to work correctly.
Avoid using the various troff preprocessors
(if you must, go ahead and use
.BR tbl (1),
but try to use the
.B IP
and
.B TP
commands instead for two-column tables).
Avoid using computations; most other tools can't process them.
Use simple commands that are easy to translate to other formats.
The following troff macros are believed to be safe (though in many cases
they will be ignored by translators):
.BR \e" ,
.BR . ,
.BR ad ,
.BR bp ,
.BR br ,
.BR ce ,
.BR de ,
.BR ds ,
.BR el ,
.BR ie ,
.BR if ,
.BR fi ,
.BR ft ,
.BR hy ,
.BR ig ,
.BR in ,
.BR na ,
.BR ne ,
.BR nf ,
.BR nh ,
.BR ps ,
.BR so ,
.BR sp ,
.BR ti ,
.BR tr .
.PP
You may also use many troff escape sequences (those sequences beginning
with \e).
When you need to include the backslash character as normal text,
use \ee.
Other sequences you may use, where x or xx are any characters and N
is any digit, include:
.BR \e\(aq ,
.BR \e\(ga ,
.BR \e- ,
.BR \e. ,
.BR \e" ,
.BR \e% ,
.BR \e*x ,
.BR \e*(xx ,
.BR \e(xx ,
.BR \e$N ,
.BR \enx ,
.BR \en(xx ,
.BR \efx ,
and
.BR \ef(xx .
Avoid using the escape sequences for drawing graphics.
.PP
Do not use the optional parameter for
.B bp
(break page).
Use only positive values for
.B sp
(vertical space).
Don't define a macro
.RB ( de )
with the same name as a macro in this or the
mdoc macro package with a different meaning; it's likely that
such redefinitions will be ignored.
Every positive indent
.RB ( in )
should be paired with a matching negative indent
(although you should be using the
.B RS
and
.B RE
macros instead).
The condition test
.RB ( if,ie )
should only have \(aqt\(aq or \(aqn\(aq as the condition.
Only translations
.RB ( tr )
that can be ignored should be used.
Font changes
.RB ( ft
and the \fB\ef\fP escape sequence)
should only have the values 1, 2, 3, 4, R, I, B, P, or CW
(the ft command may also have no parameters).
.PP
If you use capabilities beyond these, check the
results carefully on several tools.
Once you've confirmed that the additional capability is safe,
let the maintainer of this
document know about the safe command or sequence
that should be added to this list.
.SH FILES
.IR /usr/share/groff/ [*/] tmac/an.tmac
.br
.I /usr/man/whatis
.SH NOTES
.PP
By all means include full URLs (or URIs) in the text itself;
some tools such as
.BR man2html (1)
can automatically turn them into hypertext links.
You can also use the
.B UR
and
.B UE
macros to identify links to related information.
If you include URLs, use the full URL
(e.g.,
.UR http://www.kernel.org
.UE )
to ensure that tools can automatically find the URLs.
.PP
Tools processing these files should open the file and examine the first
nonwhitespace character.
A period (.) or single quote (\(aq) at the beginning
of a line indicates a troff-based file (such as man or mdoc).
A left angle bracket (<) indicates an SGML/XML-based
file (such as HTML or Docbook).
Anything else suggests simple ASCII
text (e.g., a "catman" result).
.PP
Many man pages begin with \fB\(aq\e"\fP followed by a
space and a list of characters,
indicating how the page is to be preprocessed.
For portability's sake to non-troff translators we recommend
that you avoid using anything other than
.BR tbl (1),
and Linux can detect that automatically.
However, you might want to include this information so your man page
can be handled by other (less capable) systems.
Here are the definitions of the preprocessors invoked by these characters:
.TP 3
.B e
eqn(1)
.TP
.B g
grap(1)
.TP
.B p
pic(1)
.TP
.B r
refer(1)
.TP
.B t
tbl(1)
.TP
.B v
vgrind(1)
.SH BUGS
.PP
Most of the macros describe formatting (e.g., font type and spacing) instead
of marking semantic content (e.g., this text is a reference to another page),
compared to formats like mdoc and DocBook (even HTML has more semantic
markings).
This situation makes it harder to vary the
.B man
format for different media,
to make the formatting consistent for a given media, and to automatically
insert cross-references.
By sticking to the safe subset described above, it should be easier to
automate transitioning to a different reference page format in the future.
.PP
The Sun macro
.B TX
is not implemented.
.\" .SH AUTHORS
.\" .IP \(em 3m
.\" James Clark (jjc@jclark.com) wrote the implementation of the macro package.
.\" .IP \(em
.\" Rickard E. Faith (faith@cs.unc.edu) wrote the initial version of
.\" this manual page.
.\" .IP \(em
.\" Jens Schweikhardt (schweikh@noc.fdn.de) wrote the Linux Man-Page Mini-HOWTO
.\" (which influenced this manual page).
.\" .IP \(em
.\" David A. Wheeler (dwheeler@ida.org) heavily modified this
.\" manual page, such as adding detailed information on sections and macros.
.SH SEE ALSO
.BR apropos (1),
.BR groff (1),
.BR lexgrog (1),
.BR man (1),
.BR man2html (1),
.BR whatis (1),
.BR groff_man (7),
.BR groff_www (7),
.BR man-pages (7),
.BR mdoc (7)
.SH COLOPHON
This page is part of release 5.01 of the Linux
.I man-pages
project.
A description of the project,
information about reporting bugs,
and the latest version of this page,
can be found at
\%https://www.kernel.org/doc/man\-pages/.


platforms/zcu102_arm/images/aarch64-xilinx-linux/usr/src/debug/texinfo/6.5-r0/texinfo-6.5/info/man.c
/* man.c: How to read and format man files.
   $Id: man.c 7636 2017-01-19 20:55:22Z gavin $

   Copyright 1995, 1997, 1998, 1999, 2000, 2002, 2003, 2004, 2005, 2007, 2008, 
   2009, 2011, 2012, 2013, 2014, 2015, 2016, 2017 Free Software Foundation, 
   Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

   Originally written by Brian Fox Thu May  4 09:17:52 1995. */

#include "info.h"
#include "signals.h"
#if defined (HAVE_SYS_TIME_H)
#include <sys/time.h>
#endif
#if defined (HAVE_SYS_WAIT_H)
#include <sys/wait.h>
#endif

#include "tilde.h"
#include "nodes.h"
#include "man.h"
#include "variables.h"

#if !defined (_POSIX_VERSION)
#define pid_t int
#endif

#if defined (FD_SET)
#  if defined (hpux)
#    define fd_set_cast(x) (int *)(x)
#  else
#    define fd_set_cast(x) (fd_set *)(x)
#  endif /* !hpux */
#endif /* FD_SET */

#if STRIP_DOT_EXE
static char const * const exec_extensions[] = {
  ".exe", ".com", ".bat", ".btm", ".sh", ".ksh", ".pl", ".sed", "", NULL
};
#else
static char const * const exec_extensions[] = { "", NULL };
#endif

static REFERENCE **xrefs_of_manpage (NODE *node);
static char *read_from_fd (int fd);
static char *get_manpage_contents (char *pagename);

/* We store the contents of retrieved man pages in here. */
static NODE **manpage_nodes = 0;
size_t manpage_node_index = 0;
size_t manpage_node_slots = 0;

NODE *
get_manpage_node (char *pagename)
{
  NODE *node = 0, **n, *node2 = 0;
  char *page;

  if (manpage_node_index > 0)
    for (n = manpage_nodes; (node = *n); n++)
      if (!strcmp (node->nodename, pagename))
        break;

  /* Node was not found, so we have to create it. */
  if (!node)
    {
      node = info_create_node ();
      node->fullpath = MANPAGE_FILE_BUFFER_NAME;
      node->nodename = xstrdup (pagename);
      node->flags |= N_HasTagsTable | N_IsManPage;

      /* Save this node. */
      add_pointer_to_array (node, manpage_node_index,
                            manpage_nodes,
                            manpage_node_slots, 100);
    } 

  /* Node wasn't found, or its contents were freed since last time. */
  if (!node->contents)
    {
      int plen;

      page = get_manpage_contents (pagename);
      if (!page)
        return 0;
      plen = strlen (page);

      node->contents = page;
      node->nodelen = plen;

      node->body_start = 0;
      node->references = xrefs_of_manpage (node);
      node->up = "(dir)";
    }

  node2 = xmalloc (sizeof (NODE));
  *node2 = *node;
  return node2;
}

/* Scan the list of directories in PATH looking for FILENAME.  If we find
   one that is an executable file, return it as a new string.  Otherwise,
   return a NULL pointer. */
static char *
executable_file_in_path (char *filename, char *path)
{
  struct stat finfo;
  char *temp_dirname;
  int statable, dirname_index;

  dirname_index = 0;

  while ((temp_dirname = extract_colon_unit (path, &dirname_index)))
    {
      char *temp;
      char *temp_end;
      int i;

      /* Expand a leading tilde if one is present. */
      if (*temp_dirname == '~')
        {
          char *expanded_dirname;

          expanded_dirname = tilde_expand_word (temp_dirname);
          free (temp_dirname);
          temp_dirname = expanded_dirname;
        }

      temp = xmalloc (34 + strlen (temp_dirname) + strlen (filename));
      strcpy (temp, temp_dirname);
      if (!IS_SLASH (temp[(strlen (temp)) - 1]))
        strcat (temp, "/");
      strcat (temp, filename);
      temp_end = temp + strlen (temp);

      free (temp_dirname);

      /* Look for FILENAME, possibly with any of the extensions
	 in EXEC_EXTENSIONS[].  */
      for (i = 0; exec_extensions[i]; i++)
	{
	  if (exec_extensions[i][0])
	    strcpy (temp_end, exec_extensions[i]);
	  statable = (stat (temp, &finfo) == 0);

	  /* If we have found a regular executable file, then use it. */
	  if ((statable) && (S_ISREG (finfo.st_mode)) &&
	      (access (temp, X_OK) == 0))
	    return temp;
	}

      free (temp);
    }
  return NULL;
}

/* Return the full pathname of the system man page formatter. */
static char *
find_man_formatter (void)
{
  char *man_command = getenv ("INFO_MAN_COMMAND");
  return man_command ? man_command :
                       executable_file_in_path ("man", getenv ("PATH"));
}

static char *manpage_pagename = NULL;
static char *manpage_section  = NULL;

static void
get_page_and_section (char *pagename)
{
  register int i;

  if (manpage_pagename)
    free (manpage_pagename);

  if (manpage_section)
    free (manpage_section);

  manpage_pagename = NULL;
  manpage_section  = NULL;

  for (i = 0; pagename[i] != '\0' && pagename[i] != '('; i++);

  manpage_pagename = xmalloc (1 + i);
  strncpy (manpage_pagename, pagename, i);
  manpage_pagename[i] = '\0';

  if (pagename[i] == '(')
    {
      int start;

      start = i + 1;

      for (i = start; pagename[i] != '\0' && pagename[i] != ')'; i++);

      manpage_section = xmalloc (1 + (i - start));
      strncpy (manpage_section, pagename + start, (i - start));
      manpage_section[i - start] = '\0';
    }
}

void
clean_manpage (char *manpage)
{
  mbi_iterator_t iter;
  size_t len = strlen (manpage);
  char *newpage = xmalloc (len + 1);
  char *np = newpage;
  int prev_len = 0;
  
  for (mbi_init (iter, manpage, len);
       mbi_avail (iter);
       mbi_advance (iter))
    {
      const char *cur_ptr = mbi_cur_ptr (iter);
      size_t cur_len = mb_len (mbi_cur (iter));

      if (cur_len == 1)
	{
	  if (*cur_ptr == '\b' || *cur_ptr == '\f')
	    {
	      if (np >= newpage + prev_len)
		np -= prev_len;
	    }
	  else if (ansi_escape (iter, &cur_len))
	    {
	      memcpy (np, cur_ptr, cur_len);
	      np += cur_len;
	      ITER_SETBYTES (iter, cur_len);
	    }
	  else if (show_malformed_multibyte_p || mbi_cur (iter).wc_valid)
	    *np++ = *cur_ptr;
	}
      else
	{
	  memcpy (np, cur_ptr, cur_len);
	  np += cur_len;
	}
      prev_len = cur_len;
    }
  *np = 0;
  
  strcpy (manpage, newpage);
  free (newpage);
}

static char *get_manpage_from_formatter (char *formatter_args[]);

static char *
get_manpage_contents (char *pagename)
{
  static char *formatter_args[4] = { NULL };
  char *formatted_page;

  if (formatter_args[0] == NULL)
    formatter_args[0] = find_man_formatter ();

  if (formatter_args[0] == NULL)
    return NULL;

  get_page_and_section (pagename);

  if (manpage_section)
    formatter_args[1] = manpage_section;
  else
    formatter_args[1] = "-a";

  formatter_args[2] = manpage_pagename;
  formatter_args[3] = NULL;

  formatted_page = get_manpage_from_formatter (formatter_args);

  /* If there was a section and the page wasn't found, try again
     without the section (e.g. "man 3X curses" versus "man -a curses"). */
  if (!formatted_page && manpage_section)
    {
      formatter_args[1] = "-a";
      formatted_page = get_manpage_from_formatter (formatter_args);
    }

  return formatted_page;
}

static char *
get_manpage_from_formatter (char *formatter_args[])
{
  char *formatted_page = NULL;
  int pipes[2];
  pid_t child;
  int formatter_status = 0;

  /* Open a pipe to this program, read the output, and save it away
     in FORMATTED_PAGE.  The reader end of the pipe is pipes[0]; the
     writer end is pipes[1]. */
#if PIPE_USE_FORK
  pipe (pipes);

  child = fork ();
  if (child == -1)
    return NULL;

  if (child != 0)
    {
      /* In the parent, close the writing end of the pipe, and read from
         the exec'd child. */
      close (pipes[1]);
      formatted_page = read_from_fd (pipes[0]);
      close (pipes[0]);
      wait (&formatter_status); /* Wait for child process to exit. */
    }
  else
    { /* In the child, close the read end of the pipe, make the write end
         of the pipe be stdout, and execute the man page formatter. */
      close (pipes[0]);
      freopen (NULL_DEVICE, "w", stderr);
      freopen (NULL_DEVICE, "r", stdin);
      dup2 (pipes[1], fileno (stdout));

      execv (formatter_args[0], formatter_args);

      /* If we get here, we couldn't exec, so close out the pipe and
         exit. */
      close (pipes[1]);
      exit (EXIT_SUCCESS);
    }
#else  /* !PIPE_USE_FORK */
  /* Cannot fork/exec, but can popen/pclose.  */
  {
    FILE *fpipe;
    char *cmdline;
    size_t cmdlen = 0;
    int save_stderr = dup (fileno (stderr));
    int fd_err = open (NULL_DEVICE, O_WRONLY, 0666);
    int i;

    for (i = 0; formatter_args[i]; i++)
      cmdlen += strlen (formatter_args[i]);
    /* Add-ons: 2 blanks, 2 quotes for the formatter program, 1
       terminating null character.  */
    cmdlen += 2 + 2 + 1;
    cmdline = xmalloc (cmdlen);

    if (fd_err > 2)
      dup2 (fd_err, fileno (stderr)); /* Don't print errors. */
    sprintf (cmdline, "\"%s\" %s %s",
	     formatter_args[0], formatter_args[1], formatter_args[2]);
    fpipe = popen (cmdline, "r");
    free (cmdline);
    if (fd_err > 2)
      close (fd_err);
    dup2 (save_stderr, fileno (stderr));
    if (fpipe == 0)
      return NULL;
    formatted_page = read_from_fd (fileno (fpipe));
    formatter_status = pclose (fpipe);
  }
#endif /* !PIPE_USE_FORK */

  if (!formatted_page)
    return 0;

  /* We could check the exit status of "man -a" to see if it successfully
     output a man page  However:
      * It is possible for "man -a" to output a man page and still to exit with
        a non-zero status.  This was found to happen when duplicate man pages 
        were found.
      * "man" was found to exit with a zero status on Solaris 10 even when
        it found nothing.
     Hence, treat it as a success if more than three lines were output.  (A 
     small amount of output could be error messages that were sent to standard 
     output.) */
  {
    int i;
    char *p;
    p = formatted_page;
    for (i = 0; i < 3; i++)
      {
        p = strchr (p, '\n');
        if (!p)
          {
            free (formatted_page);
            return NULL;
          }
        p++;
      }
  }

  /* If we have the page, then clean it up. */
  clean_manpage (formatted_page);

  return formatted_page;
}

/* Return pointer to bytes read from file descriptor FD.  Return value to be
   freed by caller. */
static char *
read_from_fd (int fd)
{
  struct timeval timeout;
  char *buffer = NULL;
  int bsize = 0;
  int bindex = 0;
  int select_result;
#if defined (FD_SET)
  fd_set read_fds;

  timeout.tv_sec = 15;
  timeout.tv_usec = 0;

  FD_ZERO (&read_fds);
  FD_SET (fd, &read_fds);

  select_result = select (fd + 1, fd_set_cast (&read_fds), 0, 0, &timeout);
#else /* !FD_SET */
  select_result = 1;
#endif /* !FD_SET */

  switch (select_result)
    {
    case 0:
    case -1:
      break;

    default:
      {
        int amount_read;
        int done = 0;

        while (!done)
          {
            while ((bindex + 1024) > (bsize))
              buffer = xrealloc (buffer, (bsize += 1024));
            buffer[bindex] = '\0';

            amount_read = read (fd, buffer + bindex, 1023);

            if (amount_read < 0)
              {
                done = 1;
              }
            else
              {
                bindex += amount_read;
                buffer[bindex] = '\0';
                if (amount_read == 0)
                  done = 1;
              }
          }
      }
    }

  if ((buffer != NULL) && (*buffer == '\0'))
    {
      free (buffer);
      buffer = NULL;
    }

  return buffer;
}

static REFERENCE **
xrefs_of_manpage (NODE *node)
{
  SEARCH_BINDING s;

  REFERENCE **refs = NULL;
  size_t refs_index = 0;
  size_t refs_slots = 0;
  long position;

  /* Initialize reference list to have a single null entry. */
  refs = calloc(1, sizeof (REFERENCE *));
  refs_slots = 1;

  s.buffer = node->contents;
  s.start = 0;
  s.flags = 0;
  s.end = node->nodelen;

  /* Exclude first line, which often looks like:
CAT(1)                           User Commands                          CAT(1)
  */
  s.start = strcspn (node->contents, "\n");

  /* Build a list of references.  A reference is alphabetic characters
     followed by non-whitespace text within parenthesis leading with a digit. */
  while (search_forward ("(", &s, &position) == search_success)
    {
      register int name, name_end;
      int section, section_end;

      name = position;
      if (name == 0)
        goto skip;
      else
        name--;

      /* Go to the start of a sequence of non-whitespace characters,
         checking the characters are those that should appear in a man
         page name. */
      for (; name > 0; name--)
        if (whitespace_or_newline (s.buffer[name])
            || (!isalnum (s.buffer[name])
                && s.buffer[name] != '_'
                && s.buffer[name] != '.'
                && s.buffer[name] != '-'
                && s.buffer[name] != '\033'
                && s.buffer[name] != '['))
          break;

      /* Check if reached start of buffer. */
      if (name == 0)
        goto skip;

      /* Check for invalid sequence in name. */
      if (!whitespace_or_newline (s.buffer[name]))
        goto skip;

      name++;

      if (name == position)
        goto skip; /* Whitespace immediately before '('. */

      /* 'name' is now at the start of a sequence of non-whitespace
         characters.  If we are on an ECMA-48 SGR escape sequence, skip
         past it. */
      if (s.buffer[name] == '\033' && s.buffer[name + 1] == '[')
        {
          name += 2;
          name += strspn (s.buffer + name, "0123456789;");
          if (s.buffer[name] == 'm')
            name++;
          else
            goto skip;
        }

      /* Set name_end to the end of the name, but before any SGR sequence. */
      for (name_end = name; name_end < position; name_end++)
        if (!isalnum (s.buffer[name_end])
            && s.buffer[name_end] != '_'
            && s.buffer[name_end] != '.'
            && s.buffer[name_end] != '-')
          break;

      section = position;
      section_end = 0;

      /* Look for one or two characters within the brackets, the
         first of which must be a non-zero digit and the second a letter. */
      if (!isdigit (s.buffer[section + 1])
          || s.buffer[section + 1] == '0')
        ;
      else if (!s.buffer[section + 2])
        ; /* end of buffer */
      else if (s.buffer[section + 2] == ')')
        section_end = section + 3;
      else if (!isalpha(s.buffer[section + 2]))
        ;
      else if (s.buffer[section + 3] == ')')
        section_end = section + 4;

      if (section_end)
        {
          REFERENCE *entry;
          int len = name_end - name + section_end - section;

          entry = xmalloc (sizeof (REFERENCE));
          entry->label = xcalloc (1, 1 + len);
          strncpy (entry->label, s.buffer + name, name_end - name);
          strncpy (entry->label + strlen (entry->label),
                   s.buffer + section,
                   section_end - section);

          entry->filename = xstrdup (MANPAGE_FILE_BUFFER_NAME);
          entry->nodename = xstrdup (entry->label);
          entry->line_number = 0;
          entry->start = name;
          entry->end = section_end;
          entry->type = REFERENCE_XREF;

          add_pointer_to_array (entry, refs_index, refs, refs_slots, 10);
        }

skip:
      s.start = position + 1;
    }

  return refs;
}
